libfwupdplugin/fu-block-device.c

```diff
@@ -0,0 +1,395 @@
+}
+
+/**
+ * fu_block_device_get_label:
+ * @self: a #FuBlockDevice
+ *
+ * Gets the label.
+ *
+ * Returns: string, or %NULL for unset
+ *
+ * Since: 1.7.3
+ **/
+const gchar *
+fu_block_device_get_label(FuBlockDevice *self)
+{
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_BLOCK_DEVICE(self), NULL);
+	return priv->label;
+}
+
+/**
+ * fu_block_device_set_filename:
+ * @self: a #FuBlockDevice
+ * @filename: the filename, e.g. `FWUPDATE`
+ *
+ * Sets the filename to write to the volume.
+ *
+ * Since: 1.7.3
+ **/
+void
+fu_block_device_set_filename(FuBlockDevice *self, const gchar *filename)
+{
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+
+	g_return_if_fail(FU_IS_BLOCK_DEVICE(self));
+
+	/* not changed */
+	if (g_strcmp0(priv->filename, filename) == 0)
+		return;
+
+	g_free(priv->filename);
+	priv->filename = g_strdup(filename);
+	g_object_notify(G_OBJECT(self), "filename");
+}
+
+/**
+ * fu_block_device_get_filename:
+ * @self: a #FuBlockDevice
+ *
+ * Gets the filename to write to the volume.
+ *
+ * Returns: string, or %NULL for unset
+ *
+ * Since: 1.7.3
+ **/
+const gchar *
+fu_block_device_get_filename(FuBlockDevice *self)
+{
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+	g_return_val_if_fail(FU_IS_BLOCK_DEVICE(self), NULL);
+	return priv->filename;
+}
+
+static gboolean
+fu_block_device_set_quirk_kv(FuDevice *device, const gchar *key, const gchar *value, GError **error)
+{
+	FuBlockDevice *self = FU_BLOCK_DEVICE(device);
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+
+	if (g_strcmp0(key, "BlockDeviceFilename") == 0) {
+		priv->filename = g_strdup(value);
+		return TRUE;
+	}
+	g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED, "no supported");
+	return FALSE;
+}
+
+static FuIOChannel *
+fu_block_device_io_channel_new(FuBlockDevice *self, GError **error)
+{
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+	g_autofree gchar *fn = NULL;
+
+	/* sanity check */
+	if (fu_device_get_logical_id(FU_DEVICE(self)) == NULL || priv->filename == NULL) {
+		g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "no valid path");
+		return NULL;
+	}
+	fn = g_build_filename(fu_device_get_logical_id(FU_DEVICE(self)), priv->filename, NULL);
+	return fu_io_channel_new_file(fn, error);
+}
+
+static gboolean
+fu_block_device_write_firmware(FuDevice *device,
+			       FuFirmware *firmware,
+			       FuProgress *progress,
+			       FwupdInstallFlags flags,
+			       GError **error)
+{
+	FuBlockDevice *self = FU_BLOCK_DEVICE(device);
+	FuBlockDevicePrivate *priv = GET_PRIVATE(self);
+	g_autofree gchar *fn = NULL;
+	g_autoptr(FuIOChannel) io_channel = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	/* get blob */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* sanity check */
+	if (fu_device_get_logical_id(device) == NULL || priv->filename == NULL) {
+		g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "no valid path");
+		return FALSE;
+	}
```

Comment:
same sanity check in `fu_block_device_io_channel_new`, drop it here

Reply:
God catch, fixed.